torch/csrc/jit/script/compiler.cpp

```diff
@@ -1396,17 +1397,69 @@ struct to_ir {
       const Ident& target,
       const List<Expr>& args,
       const List<Stmt>& body) {
-    // TODO: start, stop, step loop
-    if (args.size() != 1) {
+    Value *end_val, *start_val, *step_val;
+    bool isSimpleRange = (args.size() == 1);
+    if (isSimpleRange) {
+      end_val = ensureInt(range, emitExpr(args[0]));
+    } else if (args.size() == 2) {
+      start_val = ensureInt(range, emitExpr(args[0]));
+      end_val = ensureInt(range, emitExpr(args[1]));
+      step_val = end_val->owningGraph()->insertConstant(1);
+    } else if (args.size() == 3) {
+      start_val = ensureInt(range, emitExpr(args[0]));
+      end_val = ensureInt(range, emitExpr(args[1]));
+      step_val = ensureInt(range, emitExpr(args[2]));
+    } else {
       throw ErrorReport(range)
-          << "range() expects 1 argument but got " << args.size();
+          << "range() expects 1-3 arguments but got " << args.size();
     }
-    auto max_trip_count_val = ensureInt(range, emitExpr(args[0]));
     const auto& ident_name = target.name();
-    auto assigner = [ident_name, range](
+    auto assigner = [ident_name, range, start_val, step_val, isSimpleRange](
                         Value* index, std::shared_ptr<Environment> env) {
-      env->setVar(range, ident_name, index);
+      Value* derived_index;
+      if (isSimpleRange) {
+        derived_index = index;
+      } else {
+        auto g = index->owningGraph();
+        auto offset = g->insertNode(g->create(aten::mul, {index, step_val}, 1))
+                          ->output()
+                          ->setType(IntType::get());
+        derived_index =
+            g->insertNode(g->create(aten::add, {offset, start_val}, 1))
+                ->output()
+                ->setType(IntType::get());
+      }
+      env->setVar(range, ident_name, derived_index);
     };
+    Value* max_trip_count_val;
+    if (isSimpleRange) {
+      max_trip_count_val = end_val;
+    } else {
+      auto g = start_val->owningGraph();
+      // (abs(end-start) + abs(step)-1)/abs(step) = # of iterations
+      auto diff = g->insertNode(g->create(aten::sub, {end_val, start_val}, 1))
```

Comment:
Could you add a source location to all the newly created nodes? Might be worth factoring out the common bits into a function, since the pattern of insert node / create / set int output / set source location gets repeated a lot.

Reply:
Good idea.